src/stateMatrixParser.cpp

```diff
@@ -0,0 +1,443 @@
+  unsigned int number_tms=2000;
+  for (auto repeats = 50u; repeats<=max_number_tm_improve_repeats;repeats=repeats*2){
+    for (auto tm_tape_it = 500u; tm_tape_it<=max_number_tm_tape_iterations;tm_tape_it=tm_tape_it*2){
+      Metrics avg_metrics = avg_real_tm_nc_evolution(number_tms, repeats, tm_tape_it, states, alphabet_size);
+      std::cout << repeats << "\t" << tm_tape_it << "\t" << avg_metrics.amplitude << "\t" << avg_metrics.selfCompression << "\t" << avg_metrics.normalizedCompression << std::endl;
+    }
+  }
+}
+
+
+Metrics avg_real_tm_nc_evolution(unsigned int number_tms, unsigned int tm_improve_repeats , unsigned int tm_tape_iterations,unsigned int states,unsigned int alphabet_size){
+  std::vector<Metrics> vector_metrics;//also amplitude---
+  for (auto i = 0u; i<number_tms;++i){
+    vector_metrics.push_back(real_nc_evolution(tm_improve_repeats, tm_tape_iterations, states,alphabet_size, false));
+  }
+return avg_metrics(vector_metrics);
+}
+
+
+
+StateMatrix artificial_selection(std::pair<std::vector<double>,StateMatrix> state_matrix_pair_1, std::pair<std::vector<double>,StateMatrix> state_matrix_pair_2){
+   
+  
+  for(auto index=0u;index<state_matrix_pair_1.first.size();++index){
+    double absolute_value = std::abs(state_matrix_pair_1.first[index]-state_matrix_pair_2.first[index]);
+    
+    if (absolute_value>=0.1){
+      if (state_matrix_pair_2.first[index]>state_matrix_pair_1.first[index] && state_matrix_pair_2.first[index]>0){
+        
+        TuringRecord tr = state_matrix_pair_2.second.get_element(index);
+        state_matrix_pair_1.second.set_rule(index, tr);
+      }
+    }
+  }
+  return state_matrix_pair_1.second;
+}
+
+Metrics real_nc_evolution(unsigned int number_tm_it,unsigned int tm_tape_iterations, unsigned int states, unsigned int alphabet_size, bool print){
+
+  std::random_device rnd_device;
+  unsigned int seed =rnd_device();
+  Rng rng{seed};
+  //std::cerr << seed << std::endl;
+  std::vector<unsigned int> kvector{2,3,4,5,6,7,8,9};    
+  unsigned int amplitude_criteria = 50;//parameter
+  Metrics values_initial;
+  Metrics values_final;
+  BestKMarkovTables<NormalizedCompressionMarkovTable> bestMkvTableCompression(kvector, alphabet_size);
+  StateMatrix stMatrix(states, alphabet_size);
+  stMatrix.set_random(rng);
+  while(! machine_filter(stMatrix, states, alphabet_size, tm_tape_iterations, amplitude_criteria,bestMkvTableCompression) ){
+        stMatrix.set_random(rng);
+  }
+  auto vector_written_string = recreate_tm(stMatrix, states, alphabet_size, tm_tape_iterations);
+  values_initial = bestMkvTableCompression.update_string(vector_written_string);
+  auto amplitude = values_initial.amplitude;
+  auto nc = values_initial.normalizedCompression;
+  bestMkvTableCompression.reset();
+
+  auto rules = uniform_int_distribution<unsigned int>(0, (alphabet_size * states)- 1);
+  unsigned int rule_index;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The scope of the variable 'rule_index' can be reduced.](https://app.codacy.com/app/jorgeMFS/TMCompression/pullRequest?prid=4494377)